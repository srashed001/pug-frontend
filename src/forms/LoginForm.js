import {
  Button,
  Divider,
  Stack,
  TextField,
  Typography,
  Box,
  Alert,
} from "@mui/material";
import { useState } from "react";
import { Controller, useForm } from "react-hook-form";
import { useNavigate } from "react-router-dom";

import ArrowBackIosIcon from "@mui/icons-material/ArrowBackIos";
import "./login.css";
import "../homepage/publicHomepage.css";



const inputOptions = {
  width: {
    xs: "100%",
  },
  marginBottom: 2,
  fontSize: 10,
  color: "black",
};

function LoginForm({ login }) {
  let navigate = useNavigate();

  const {
    reset,

    control,
    handleSubmit,
    formState: { errors },
  } = useForm({
    defaultValues: {
      username: "",
      password: "",
    },
  });

  const [formErrors, setFormErrors] = useState(null);

  async function handleLogin(data) {
    let res = await login(data);
    if (res.success) {
      navigate(`/`);
    } else {
      setFormErrors("Invalid username/password");
      reset();
    }
  }

  function goHome() {
    navigate(`/`);
  }

  const handleSignup = () => {
    navigate("/signup");
  };

  return (

      <Stack className="publicHomepage" sx={{ alignItems: "center" }}>
        <Box
          sx={{
            display: "inline-flex",
            backgroundColor: "#E5383B",
            position: "fixed",
            top: "3.5rem",
            zIndex: "10",
            width: "100%",
            boxShadow: 3,
          }}
        >
          <Box sx={{ display: "flex", alignItems: "center" }}>
            <Button
              sx={{ color: "#FFFFFF", marginX: 1 }}
              startIcon={<ArrowBackIosIcon />}
              onClick={goHome}
            >
              Home
            </Button>
          </Box>
          <Box>
            <Typography sx={{ fontSize: 20, padding: 1, color: "#FFFFFF" }}>
              Login
            </Typography>
          </Box>
        </Box>
        {formErrors && (
          <Box sx={{ position: "absolute", top: "6rem", width: "100%" }}>
            <Alert severity="error" onClose={() => setFormErrors(null)}>
              {formErrors}
            </Alert>
          </Box>
        )}
        <Stack
          sx={{
            boxShadow: 5,
            padding: 1,
            paddingY: 3,
            margin: 1,
            marginTop: 14,
            borderRadius: 2,
            backgroundColor: "#FFFFFF",
            width: { xs: 300 },
          }}
        >
          <Stack component="form" onSubmit={handleSubmit(handleLogin)}>
            <Controller
              name="username"
              control={control}
              rules={{ required: "username required" }}
              render={({ field }) => (
                <TextField
                  {...field}
                  size="small"
                  aria-label="username"
                  sx={inputOptions}
                  error={!!errors.username}
                  helperText={
                    errors.username ? errors.username.message : "username"
                  }
                />
              )}
            />
            <Controller
              name="password"
              control={control}
              rules={{ required: "password required" }}
              render={({ field }) => (
                <TextField
                  {...field}
                  size="small"
                  type="password"
                  aria-label="password"
                  sx={inputOptions}
                  error={!!errors.password}
                  helperText={
                    errors.password ? errors.password.message : "password"
                  }
                />
              )}
            />
            <Button type="submit" variant="contained">
              Login
            </Button>
            <Typography
              sx={{ textAlign: "center", marginY: 1 }}
              component={Box}
            >
              Forgot Password?
            </Typography>
          </Stack>
          <Divider />
          <Button onClick={handleSignup}>Create new account</Button>
        </Stack>
      </Stack>

  );
}

export default LoginForm;



