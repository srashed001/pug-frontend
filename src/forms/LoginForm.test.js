import * as React from 'react'
import {render, screen, act} from '../test/test-utils'
import userEvent from '@testing-library/user-event'
import LoginForm from './LoginForm'



test('it renders without crashing', () => {
    const handleSubmit = jest.fn()
    render(<LoginForm login={handleSubmit} />)
})

test('it matches snapshot', () => {
    const handleSubmit = jest.fn()
    const {asFragment } = render(<LoginForm login={handleSubmit} />)
    expect(asFragment()).toMatchSnapshot()
 
})

test('it submits form data when filled out', async () => {

    const handleSubmit = jest.fn()
    // const {debug} = render(<LoginForm login={handleSubmit} />)
    const [usernameInput, passwordInput] = screen.getAllByDisplayValue("")
    const submitButton = screen.getByRole('button', {name: /login/i})


    
    const username = 'test1'
    const password = 'password1'
    
    await userEvent.type(usernameInput, username)
    await userEvent.type(passwordInput, password)
    
    // debug()
    await userEvent.click(submitButton)
  
    expect(handleSubmit).toHaveBeenCalledWith({
      username,
      password,
    })
    expect(handleSubmit).toHaveBeenCalledTimes(1)
})


