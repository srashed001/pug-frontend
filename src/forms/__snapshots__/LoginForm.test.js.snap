// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="publicHomepage css-163ljxj-MuiStack-root"
  >
    <div
      class="MuiBox-root css-12yy1tz"
    >
      <div
        class="MuiBox-root css-70qvj9"
      >
        <button
          class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-6inqtj-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowBackIosIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11.67 3.87 9.9 2.1 0 12l9.9 9.9 1.77-1.77L3.54 12z"
              />
            </svg>
          </span>
          Home
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiBox-root css-0"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-19wt13x-MuiTypography-root"
        >
          Login
        </p>
      </div>
    </div>
    <div
      class="css-1bg9mof-MuiStack-root"
    >
      <form
        class="css-nen11g-MuiStack-root"
      >
        <div
          aria-label="username"
          class="MuiFormControl-root MuiTextField-root css-zyp2ld-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-describedby=":r2:-helper-text"
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r2:"
              name="username"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-hdw1oc"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeSmall MuiFormHelperText-contained css-k4qjio-MuiFormHelperText-root"
            id=":r2:-helper-text"
          >
            username
          </p>
        </div>
        <div
          aria-label="password"
          class="MuiFormControl-root MuiTextField-root css-zyp2ld-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-describedby=":r3:-helper-text"
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r3:"
              name="password"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-hdw1oc"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeSmall MuiFormHelperText-contained css-k4qjio-MuiFormHelperText-root"
            id=":r3:-helper-text"
          >
            password
          </p>
        </div>
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Login
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <div
          class="MuiTypography-root MuiTypography-body1 MuiBox-root css-18x9swi-MuiTypography-root"
        >
          Forgot Password?
        </div>
      </form>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <button
        class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Create new account
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
